drivers/SmartThings/deepsmart/src/deepsmart/wisers.lua

```diff
@@ -0,0 +1,825 @@
+      local wiser_port = wiser_info.port
+      local wiser_ip = wiser_info.ip
+      if (wiser_port == nil) then
+        wiser_port = 8000
+      end
+      if (wiser_ip == nil) then
+        wiser_ip = ''
+      end
+      local info = {
+        ip = wiser_ip,
+        uuid = wiser_index_code,
+        port = wiser_port
+      }
+      wisers[idx] = info
+      idx = idx + 1
+    end
+  end
+  log.info('get wisers count '..#wisers)
+  Wisers.add_wisers(wisers, true)
+  -- init all wisers
+  for _,wiser in pairs(Wisers.wisers) do
+    local wiser_index_code = wiser.wiser_index_code
+    local configstr = driver.datastore[wiser_index_code].config or ''
+    local dp2knxstr = driver.datastore[wiser_index_code].dp2knx or ''
+    local dpenumstr = driver.datastore[wiser_index_code].dpenum or ''
+    local add_devs = {}
+    local del_devs = {}
+    log.info('load wiser '..wiser.wiser_index_code..' config '..configstr)
+    wiser["config"] = Devices.load_config(wiser.indexCode, configstr, nil, add_devs, del_devs)
+    -- get dp info
+    log.info('load wiser '..wiser.wiser_index_code..' dp2knx '..dp2knxstr)
+    wiser["dp2knx"] = dp2knx.load_config(wiser.indexCode, dp2knxstr)
+    -- get enum info
+    log.info('load wiser '..wiser.wiser_index_code..' dpenum '..dpenumstr)
+    wiser["dpenum"] = dpenum.load_config(wiser.indexCode, dpenumstr)
+  end
+  -- save device_network_id->id
+  local devices = driver:get_devices()
+  for i,v in pairs(devices) do
+    if (v.parent_assigned_child_key ~= nil) then
+      log.debug('save dev '..i..' id '..v.id..' networkid '..v.parent_assigned_child_key)
+      Wisers.idmap[v.parent_assigned_child_key] = v.id
+    end
+  end
+  -- start ssdp coroutine to refresh wiser info
+  driver:call_on_schedule(
+  30,
+  function ()
+    return Wisers.check_ssdp(discovery)
+  end,
+  'Refresh schedule')
+  return true,nil
+end
+------------------
+-- add wisers
+-- create wiser client(https)
+-- api: load device configs, dp2knx config, dpenum config
+------------------
+function Wisers.add_wisers(wisers, is_add)
+  -- init all wisers
+  for i,v in pairs(wisers) do
+    log.info('add wiser '..v.uuid..' ip:port('..v.ip..':'..v.port..')')
+    local wiser = Wisers.get_wiser(v.uuid)
+    if (not is_add and wiser == nil) then
+      log.trace('ignore new wiser '..v.uuid)
+      goto continue
+    end
+    -- first check wiser ip, if ip is same then do nothing
+    if (wiser ~= nil and wiser.ip == v.ip) then
+      log.trace('wiser '..v.uuid..' is same')
+      goto continue
+    end
+    -- if wiser is wiser then create new wiser
+    if (wiser == nil) then
+      Wisers.wisers[v.uuid] = {}
+      wiser = Wisers.wisers[v.uuid]
+      wiser["dp2knx"] = dp2knx.load_config(v.uuid, nil)
+      wiser["dpenum"] = dpenum.load_config(v.uuid, nil)
+      wiser.state = 1
+      wiser.ctrlmap = {}
+      Wisers.driver:call_on_schedule(
+      2,
+      function ()
+        return Wisers.wiser_loop(v.uuid)
+      end,
+      'wiser loop')
```

Comment:
@refuseRed I am trying to understand whether the 2 second polling is something we are forced to do because that is how the devices are designed, or if it was a design choice.

Reply:
wiser loop will poll updated devices by given time every 2 seconds. wiser loop will poll updated devices by given time every 2 seconds.

Reply:
Is the approach to poll the wisers/bridges something that is required by the wisers, or is it a design choice you made? Does there exist any other mechanism for having the device tell the driver of device updates without the driver having to poll for them (i.e. websocket, sse, other listening server setup, etc)?

Reply:
Hi @cjswedes 

> Is the approach to poll the wisers/bridges something that is required by the wisers, or is it a design choice you made?

I thought it's a design choice. I'll try to explain the background clearly in my reply below. 

> Does there exist any other mechanism for having the device tell the driver of device updates without the driver having to poll for them (i.e. websocket, sse, other listening server setup, etc)?

DeepSmart bridge supports `http/https/websocket/socket`

We (SRC-N and HQ) did some simple codes review before DeepSmart submitted this PR.
In the code version before the PR submission, DeepSmart use `http` and `socket` to communicate with their bridge.

We thought it's not safe so we required DeepSmart to use TLS to encrypt communication.
Since their bridge doesn't support WWS, so `https` is the only choice. 

But replacing `socket` with `https` introduced a problem that bridged devices can't update their status to ST Station actively. To solve this problem, DeepSmart team added an feature to log all updated devices in DeepSmart bridge.
The updated devices list is exposed through an https url.

Changed Dev <--- Report to DeepSmart Bridge ---> DeepSmart <--- HTTPS URL (Changed dev list) ---> ST Hub

So ST HUB needs to query changed dev list actively. That is why we need to run  `Wisers.wiser_loop` periodically.

Reply:
Thank you for the context. Its a shame to do a polling approach when something else exists, but since WWS isn't available on the device and the tls encryption is required, I understand why you are doing the polling approach.

Ill admit that its weird wws isn't supported by the devices since the tls handshake is supported; my understanding is that wws just requires the tls handshake prior to doing the websocket handshake. I know the hub supports it, and sonos is an example of our use of it. ¯\\_(ツ)_/¯

Reply:
Hi, @cjswedes 

Thanks for your quick response. 

> Ill admit that its weird wws isn't supported by the devices since the tls handshake is supported; my understanding is that wws just requires the tls handshake prior to doing the websocket handshake.

Yes. I'll confirm it with DeepSmart asap. 

In the previous discussion with DeepSmart, we saw that only the above 4 communications methods `(http/https/websocket/socket)` were supported, so we asked DeepSmart to change `socket` to `https` directly.
We have no further confirmation regarding `WWS` support issues. This is indeed an omission.

However, even if DeepSmart Bridge truly supports the feature, it might be too late to make changes according to the current schedule. If possible, could we consider iterating on this in the future ?

CC @lelandblue

Reply:
> However, even if DeepSmart Bridge truly supports the feature, it might be too late to make changes according to the current schedule. If possible, could we consider iterating on this in the future ?

I would be fine with iterating on it and providing secure websocket support for manual state updates in the future.

Reply:
> > However, even if DeepSmart Bridge truly supports the feature, it might be too late to make changes according to the current schedule. If possible, could we consider iterating on this in the future ?
> 
> I would be fine with iterating on it and providing secure websocket support for manual state updates in the future.

Hi @cjswedes 

I've confimed with DeepSmart that their Bridge does not support `Websocket over TLS`. 
The library they currently use for firmware development does not support it.